server/mux.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/getkin/kin-openapi/pathpattern"
+
+	ac "github.com/avenga/couper/accesscontrol"
+	"github.com/avenga/couper/config/request"
+	"github.com/avenga/couper/config/runtime"
+	"github.com/avenga/couper/errors"
+	"github.com/avenga/couper/handler"
+	"github.com/avenga/couper/utils"
+)
+
+// Mux is a http request router and dispatches requests
+// to their corresponding http handlers.
+type Mux struct {
+	apiPath        string
+	apiErrHandler  *errors.Template
+	endpointRoot   *pathpattern.Node
+	fileBasePath   string
+	fileHandler    http.Handler
+	fileErrHandler *errors.Template
+	router         *openapi3filter.Router
+	spaRoot        *pathpattern.Node
+}
+
+var allowedMethods = []string{
+	http.MethodGet,
+	http.MethodHead,
+	http.MethodPost,
+	http.MethodPut,
+	http.MethodPatch,
+	http.MethodDelete,
+	http.MethodOptions,
+}
+
+func NewMux(options *runtime.MuxOptions) *Mux {
+	opts := options
+	if opts == nil {
+		opts = &runtime.MuxOptions{
+			APIErrTpl:  errors.DefaultJSON,
+			FileErrTpl: errors.DefaultHTML,
+		}
+	}
+
+	mux := &Mux{
+		apiPath:        opts.APIPath,
+		apiErrHandler:  opts.APIErrTpl,
+		fileBasePath:   opts.FileBasePath,
+		fileHandler:    opts.FileHandler,
+		fileErrHandler: opts.FileErrTpl,
+		endpointRoot:   &pathpattern.Node{},
+		spaRoot:        &pathpattern.Node{},
+	}
+
+	for path, h := range opts.EndpointRoutes {
+		// TODO: handle method option per endpoint configuration
+		mux.mustAddRoute(mux.endpointRoot, allowedMethods, path, h)
+	}
+
+	for path, h := range opts.SPARoutes {
+		mux.mustAddRoute(mux.spaRoot, []string{http.MethodGet}, path, h)
+	}
+
+	return mux
+}
+
+func (m *Mux) MustAddRoute(method, path string, handler http.Handler) *Mux {
+	methods := allowedMethods[:]
+	if method != "*" {
+		um := strings.ToUpper(method)
+		var allowed bool
+		for _, am := range allowedMethods {
+			if um == am {
+				allowed = true
+				break
+			}
+		}
+		if !allowed {
+			panic(fmt.Errorf("method not allowed: %q, path: %q", um, path))
+		}
+
+		methods = []string{um}
+	}
+	return m.mustAddRoute(m.endpointRoot, methods, path, handler)
+}
+
+func (m *Mux) mustAddRoute(root *pathpattern.Node, methods []string, path string, handler http.Handler) *Mux {
+	const wildcardReplacement = "/{_couper_wildcardMatch*}"
+	const wildcardSearch = "/**"
+
+	for _, method := range methods {
+		pathOptions := &pathpattern.Options{}
+
+		if strings.HasSuffix(path, wildcardSearch) {
+			pathOptions.SupportWildcard = true
+			path = path[:len(path)-len(wildcardSearch)] + wildcardReplacement
+		}
+
+		node, err := root.CreateNode(method+" "+path, pathOptions)
+		if err != nil {
+			panic(fmt.Errorf("create path node failed: %s %q: %v", method, path, err))
+		}
+
+		node.Value = &openapi3filter.Route{
+			Method:  method,
+			Path:    path,
+			Handler: handler,
+		}
+	}
+	return m
+}
+
+func (m *Mux) FindHandler(req *http.Request) http.Handler {
+	var route *openapi3filter.Route
+	node, paramValues := m.match(m.endpointRoot, req)
```

Comment:
Hier hat sich das Verhalten geändert, vor allem wenn der API-Path dem FS-Path oder SPA-Path entspricht (clash). Das Verhalten sollte sich an die gelöschte `server/handler.go` `Match()` orientieren.

Reply:
You are talking about the [behaviour below](https://github.com/avenga/couper/blob/0403e1f41013b5d8f320dd1eb767b7e9369d0aa9/server/mux.go#L123-L146) which seems to be exactly the one from the replaced [`Match()` method](https://github.com/avenga/couper/blob/3d93a697a493ac5a9bdecca78cd41a35254351d2/server/handler.go#L24).

1. Lookup existing api routes
2. Ensure we still have matched an api base path and return an api error
3. Otherwise ask the fileHandler for an possible response, unwrap protected handler if required*
4. No fileHandler is there or have no response ask the spa paths
5. No matching spa path but a fileHandler was defined -> 404
6. Nothing at all results in a configuration error

*As I have already noticed the fileHandler setup needs a fix to consider hosts mapping too. I am working on it.

API-Path wins over same existing FS-Path or SPA-Path since the api root node gets asked first!? Additionally the API -vs- Spa-Path should get solved at setup time.